README.MD
ARed - Redis Cluster based on Rendezvous Hashing (HRW).

WARNING! Highly experimental. Use at your own risk

USAGE:

See tests for examples. More details soon.



CHANGELOG.md
# 2018-01-13, v0.2.2
* Added option in rebalancer to flush an instance that has dropped out. Read more in README.md
* Improved rebalancer testing
* Updated license 
* Updated README.md

# 2018-01-06, v0.2.1
* Did some micro optimizations to improve performance a little
* Updated benchmark results to reflect code changes and bench setup
* Improved code consistency
* Updated README.md

# 2018-01-05, v0.2.0
* Added benchmark
* Added ability to change separator for error and result objects
* Updated README.md

# 2017-12-19, v0.1.9
* Fixed Redis error in test
* Fixed Rebalancer to support replication
* Updated README.md

# 2017-12-17, v0.1.8
* Changed Redis error from JSON.stringify to toString for better readability

# 2017-12-15, v0.1.7
* Rewrote error logging system and tests

# 2017-12-15, v0.1.6
* Made sure `Helper.flatten` doesn't ignore empty arrays

# 2017-12-15, v0.1.5
* Added naive `mget` command implementation and tests

# 2017-12-14, v0.1.4
* Bumped version to correct number

# 2017-12-14, v0.1.3
* Fixed CHANGELOG.md

# 2017-12-14, v0.1.2
* Fixed the npm package name

# 2017-12-14, v0.1.1
* Added git repo link in package.json

# 2017-12-14, v0.1.0
* Renamed package ARed to ared
* Updated README.md
* Released NPM package
* Made sure all caps commands are accepted

# 2017-12-12, v0.0.7
* Fixed issue when all values are `null` in multi key commands 
* Added test case for all `null` values in multi keys commands

# 2017-12-12, v0.0.6
* Fixed issue with `null` values in multi key commands 
* Added test case for `null` values in multi keys commands

# 2017-12-12, v0.0.5
* Exposed `Helper` class for external access
* Renamed internal method `Helper.shuffle` to follow style guidelines

# 2017-12-12, v0.0.4
* Added extra keyword for package.json
* Bumped version in package.json
* Updated example in README.md
* Revamped error logging system to have more details
* Added support for multi key commands (only PFCOUNT for now)
* Temporary removed `mget` from read commands
* Added `dump` support for proper ARead random reads
* Updated tests

# 2017-11-19, v0.0.3
* Added example to README.md
* Made sure library does not depend on globals

# 2017-11-19, v0.0.2

* Added CHANGELOG.md
* Added LICENSE
* Added `hgetall`, `pfcount`, `smembers` support for proper ARead random reads
* Renamed README.MD to README.md
* Formatted README.md for better readability
* Updated README.md with some extra info

test.js
function traverse(obj) {
    if (typeof obj === "object") {
        for (let key in obj) {
            if (obj.hasOwnProperty(key)) {
                traverse(obj[key])
            }
        }
    } else {
        return obj
    }
}

var a = {b: {k: "error3", c: {d: null, e: "error2"}, f: null, g: "error2"}}


function findById(obj) {
    let result = null

    for (let p in obj) {
        if (typeof obj[p] !== "object") {
            return obj[p]
        } else {
            if (typeof obj[p] === "object") {
                result = findById(obj[p])

                if (result) {
                    return result
                }
            }
        }
    }

    return result
}

function flatten(ob) {
    var toReturn = {};

    for (var i in ob) {
        if (!ob.hasOwnProperty(i)) continue;

        if ((typeof ob[i]) == 'object') {
            var flatObject = flatten(ob[i]);
            for (var x in flatObject) {
                if (!flatObject.hasOwnProperty(x)) continue;

                toReturn[i + '.' + x] = flatObject[x];
            }
        } else {
            toReturn[i] = ob[i];
        }
    }
    return toReturn;
};

console.log(flatten(a))


tests/helper.js
"use strict"

global.__base = __dirname + "/../"

const should = require("should") // eslint-disable-line no-unused-vars

const Helper = require(`${__base}libs/helper`)

describe("HELPER", () => {
    const redisOptions = {
        r1: {
            host: "127.0.0.1",
            port: 6379,
            enable_offline_queue: false // eslint-disable-line camelcase
        },
        r2: {
            host: "127.0.0.1",
            port: 6380,
            enable_offline_queue: false // eslint-disable-line camelcase
        },
        r3: {
            host: "127.0.0.1",
            port: 6381,
            enable_offline_queue: false // eslint-disable-line camelcase
        }
    }

    it("Should get the highest weight option for key 'foo'", (done) => {
        const key = "foo"

        const clients = Helper.getClients(redisOptions, key, 1)

        clients[0][0].should.be.equal("r3")
        clients[0][1].should.be.equal("15fa4405d23e2087")

        done()
    })

    it("Should get 2 highest weight and randomized options for key 'foo'", (done) => {
        const key = "foo"

        let first = Helper.getClients(redisOptions, key, 2)

        const get = () => {
            const clients = Helper.getClients(redisOptions, key, 2)

            if (first[0][1] !== clients[0][1]) {
                first = clients

                done()
            } else {
                get()
            }
        }

        get()
    })

    it("Should flatten a deep object to one dimensional key => value object", (done) => {
        const object = {
            a: {b: null},
            c: {
                d: "foo",
                e: {f: "bar"}
            },
            g: []
        }

        const flatObject = Helper.flatten(object, ".")

        ;(flatObject["a.b"] === null).should.be.true()
        flatObject["c.d"].should.be.equal("foo")
        flatObject["c.e.f"].should.be.equal("bar")
        flatObject["g"].length.should.be.equal(0)

        done()
    })
})

tests/rebalancer.js
"use strict"

global.__base = __dirname + "/../"

const redis = require("redis")
const should = require("should") // eslint-disable-line no-unused-vars

const ARed = require(`${__base}libs/index`)

const Rebalancer = require(`${__base}tools/rebalancer`)

describe("REBALANCER", () => {
    const redisOptions = {
        r1: {
            host: "127.0.0.1",
            port: 6379
        },
        r2: {
            host: "127.0.0.1",
            port: 6380
        },
        r3: {
            host: "127.0.0.1",
            port: 6381
        },
        r4: {
            host: "127.0.0.1",
            port: 6382,
            flush: true
        }
    }

    const total = 1200
    const replication = 2
    const redisClients = {}
    const keys = []
    let rebalancer = null

    before((done) => {
        const populate = (iteration, callback) => {
            let count = total

            for (let i = 0; i < total; i++) {
                const key = `test_${i}`
                const value = i * iteration

                keys.push(key)

                ared.exec("set", [key, value], (error) => {
                    if (error) {
                        throw error
                    }

                    if (--count === 0) {
                        return callback()
                    }
                })
            }
        }

        for (let id in redisOptions) {
            redisClients[id] = redis.createClient(redisOptions[id])

            redisClients[id].send_command("FLUSHALL")
        }

        let ared = new ARed()

        ared.replication = replication

        ared.listen(null, redisOptions, null, () => {
            // Populate when all nodes are present
            populate(1, () => {
                const dropoutClient = redisOptions["r4"]

                delete redisOptions["r4"]

                ared = new ARed()

                ared.replication = replication
                ared.writePolicy = replication

                ared.listen(null, redisOptions, null, () => {
                    // Populate when one node has dropped out
                    populate(2, () => {
                        redisOptions["r4"] = dropoutClient

                        ared = new ARed()

                        ared.replication = replication
                        ared.writePolicy = replication

                        ared.listen(null, redisOptions, null, () => {
                            // Rebalance with all nodes present again
                            rebalancer = new Rebalancer(redisOptions)

                            rebalancer.replication = replication
                            ared.writePolicy = replication

                            done()
                        })
                    })
                })
            })
        })
    })

    it("Simulate a node drop out and it's re-add", (done) => {
        rebalancer.start(() => {
            const results = {}
            const values = {}

            let x = Object.keys(redisClients).length

            for (let clientId in redisClients) {
                results[clientId] = {}
                values[clientId] = {}

                redisClients[clientId].send_command("DBSIZE", (error, result) => {
                    if (error) {
                        throw error
                    }

                    results[clientId] = result

                    redisClients[clientId].send_command("MGET", keys, (error, result2) => {
                        if (error) {
                            throw error
                        }

                        for (let i = 0; i < keys.length; i++) {
                            if (result2[i]) {
                                values[clientId][keys[i]] = result2[i]
                            }
                        }

                        if (--x === 0) {
                            for (let clientId in results) {
                                switch (clientId) {
                                    case "r1": {
                                        results[clientId].should.be.equal(587)
                                        break
                                    }
                                    case "r2": {
                                        results[clientId].should.be.equal(592)
                                        break
                                    }
                                    case "r3": {
                                        results[clientId].should.be.equal(603)
                                        break
                                    }
                                    case "r4": {
                                        results[clientId].should.be.equal(618)
                                        break
                                    }
                                }
                            }

                            const check = {}
                            let counter = 0

                            for (let clientId in values) {
                                for (let key in values[clientId]) {
                                    if (typeof check[key] === "undefined") {
                                        check[key] = values[clientId][key]

                                        counter++
                                    } else {
                                        check[key].should.be.equal(values[clientId][key])

                                        counter++
                                    }
                                }
                            }

                            counter.should.be.equal(total * replication)

                            done()
                        }
                    })
                })
            }
        })
    })
})

libs/index.js
"use strict"

const http = require("http")

const redis = require("redis")

const Commands = require("./commands")
const Helper = require("./helper")

class ARed extends Commands {
    constructor() {
        super()

        this.replication = 1 // 1 - no replication, 2..n - replication factor
        this.writePolicy = 1 // 0 - don't wait at all, 1..n - wait for x replications
        this.debug = false // if true, will return full path of the key (same as errors)
        this.separator = "." // Reserved for error and result flattening. Change this if your keys have dots in them.

        this._agent = new http.Agent({keepAlive: true})

        this._server = null
        this._clients = {}

        this.redisErrors = {}
    }

    listen(serverOptions, redisOptions, forwardingOptions, callback) {
        if (!serverOptions && !redisOptions && !forwardingOptions) {
            throw new Error("You need to supply at least serverOptions, redisOptions or forwardingOptions.")
        }

        let x = 0

        if (forwardingOptions) {
            x += Object.keys(forwardingOptions).length

            for (let id in forwardingOptions) {
                this._clients[id] = forwardingOptions[id]

                if (--x === 0) {
                    if (callback) {
                        return callback()
                    }
                }
            }
        }

        if (redisOptions) {
            x += Object.keys(redisOptions).length

            for (let id in redisOptions) {
                const client = redis.createClient(redisOptions[id])

                client.on("error", (error) => {
                    this.redisErrors[id] = error
                })

                client.on("ready", () => {
                    delete this.redisErrors[id]

                    if (--x === 0) {
                        if (callback) {
                            return callback()
                        }
                    }
                })

                this._clients[id] = client
            }
        }

        if (serverOptions) {
            x++

            this._server = http.createServer((req, res) => {
                let body = ""

                req.on("data", (data) => {
                    body += data
                })

                req.on("end", () => {
                    const parsed = JSON.parse(body)

                    this.exec(parsed[0], parsed[1], (error, result) => {
                        res.end(JSON.stringify([error, result]))
                    }, false)
                })
            })

            this._server.on("error", (error) => {
                throw error
            })

            this._server.listen(serverOptions, () => {
                if (--x === 0) {
                    if (callback) {
                        return callback()
                    }
                }
            })
        }
    }

    exec(command, args, callback, firstCall = true) {
        command = command.toLowerCase()

        if (typeof this[command] !== "undefined") {
            this[command](args, callback)
        } else {
            if (Array.isArray(args[0]) === true) {
                throw new Error(`Command ${command} is currently not supported with multi keys.`)
            }

            this._scatter(command, args, this.writePolicy, (errors, results) => {
                if (firstCall === true) {
                    errors = Helper.flatten(errors, this.separator)

                    if (Object.keys(errors).length === 0) {
                        errors = null
                    }

                    results = Helper.flatten(results, this.separator)

                    let preparedResults = {}

                    if (this.debug === false) {
                        for (let path in results) {
                            preparedResults[path.split(this.separator)[0]] = results[path]
                        }
                    } else {
                        preparedResults = results
                    }

                    return callback(errors, preparedResults)
                } else {
                    return callback(errors, results)
                }
            })
        }
    }

    _scatter(command, args, writePolicy, callback) {
        const keys = (Array.isArray(args[0]) === true) ? args[0] : [args[0]]

        const isRead = this._readCommands[command] || false

        const errors = {}
        const results = {}
        let x = keys.length

        for (let i = 0; i < keys.length; i++) {
            args[0] = keys[i]

            results[keys[i]] = {}

            const clients = Helper.getClients(this._clients, keys[i], this.replication)

            if (isRead === true) {
                this._send(clients, 0, isRead, command, args, (error, result, clientId) => {
                    if (error) {
                        if (typeof errors[keys[i]] === "undefined") {
                            errors[keys[i]] = {}
                        }

                        errors[keys[i]][clients[clientId][0]] = error
                    }

                    results[keys[i]][clients[clientId][0]] = result

                    if (--x === 0) {
                        return callback(errors, results)
                    }
                })
            } else {
                if (writePolicy === 0) {
                    for (let j = 0; j < clients.length; j++) {
                        this._send(clients, j, isRead, command, args)
                    }
                } else {
                    let y = writePolicy

                    for (let j = 0; j < clients.length; j++) {
                        this._send(clients, j, isRead, command, args, (error, result, clientId) => {
                            if (error) {
                                if (typeof errors[keys[i]] === "undefined") {
                                    errors[keys[i]] = {}
                                }

                                errors[keys[i]][clients[clientId][0]] = error
                            }

                            results[keys[i]][clients[clientId][0]] = result

                            if (--y === 0) {
                                if (--x === 0) {
                                    return callback(errors, results)
                                }
                            }
                        })
                    }
                }
            }
        }
    }

    _send(clients, clientId, isRead, command, args, callback) {
        const client = this._clients[clients[clientId][0]]

        if (client.constructor.name === "RedisClient") {
            client.send_command(command, args, (error, result) => {
                if (error) {
                    error = error.toString()
                }

                if (isRead === true) {
                    const nextClientId = clientId + 1

                    if (error && nextClientId < this.replication) {
                        this._send(clients, nextClientId, isRead, command, args, callback)
                    } else {
                        return callback(error, result, clientId)
                    }
                } else if (callback) {
                    return callback(error, result, clientId)
                }
            })
        } else {
            const postData = JSON.stringify([command, args])

            const options = {
                host: client.host,
                port: client.port,
                method: "POST",
                agent: this._agent,
                headers: {
                    "Content-Type": "application/json",
                    "Content-Length": Buffer.byteLength(postData)
                }
            }

            const req = http.request(options, (res) => {
                let body = ""

                res.setEncoding("utf8")

                res.on("data", (chunk) => {
                    body += chunk
                })

                res.on("end", () => {
                    const parsed = JSON.parse(body)

                    if (isRead === true) {
                        const nextClientId = clientId + 1

                        const error = Helper.flatten(parsed[0], this.separator)

                        let errorFound = false

                        for (let path in error) {
                            if (error[path]) {
                                errorFound = true
                            }
                        }

                        if (errorFound === true && nextClientId < this.replication) {
                            this._send(clients, nextClientId, isRead, command, args, callback)
                        } else {
                            return callback(parsed[0], parsed[1], clientId)
                        }
                    } else if (callback) {
                        return callback(parsed[0], parsed[1], clientId)
                    }
                })
            })

            req.on("error", (error) => {
                if (error) {
                    error = JSON.stringify(error)
                }

                if (isRead === true) {
                    const nextClientId = clientId + 1

                    if (error && nextClientId < this.replication) {
                        this._send(clients, nextClientId, isRead, command, args, callback)
                    } else {
                        return callback(error, null, clientId)
                    }
                } else if (callback) {
                    return callback(error, null, clientId)
                }
            })

            req.end(postData)
        }
    }
}

module.exports = ARed

libs/helper.js
"use strict"

const murmurhash = require("murmurhash-native")

class Helper {
    static getClients(clients, key, replication) {
        let selectedClients = []

        for (let id in clients) {
            selectedClients.push([id, murmurhash.murmurHash64(id + key)])
        }

        selectedClients = selectedClients.sort(Helper._sort)

        return Helper._shuffle(selectedClients.slice(0, replication))
    }

    static _sort(a, b) {
        return (a[1] > b[1]) - (a[1] < b[1])
    }

    static _shuffle(range) {
        let counter = range.length

        while (counter > 0) {
            let index = Math.floor(Math.random() * counter)

            counter--

            let temp = range[counter]

            range[counter] = range[index]
            range[index] = temp
        }

        return range
    }

    static flatten(obj, separator) {
        const toReturn = {}

        for (let key in obj) {
            if (typeof obj[key] === "object" && obj[key] !== null && Array.isArray(obj[key]) === false
                && Buffer.isBuffer(obj[key]) === false) {

                const flatObject = Helper.flatten(obj[key], separator)

                for (let x in flatObject) {
                    toReturn[key + separator + x] = flatObject[x]
                }
            } else {
                toReturn[key] = obj[key]
            }
        }

        return toReturn
    }
}

module.exports = Helper

libs/commands.js
"use strict"

const Helper = require("./helper")

class Commands {
    constructor() {
        this.prefix = "ared"

        this._readCommands = {
            bitcount: true,
            bitpos: true,
            get: true,
            mget: true,
            getbit: true,
            getrange: true,
            strlen: true,
            zcard: true,
            zcount: true,
            zlexcount: true,
            zrange: true,
            zrangebylex: true,
            zrevrangebylex: true,
            zrangebyscore: true,
            zrank: true,
            zrevrange: true,
            zrevrangebyscore: true,
            zrevrank: true,
            zscore: true,
            zscan: true,
            scard: true,
            smembers: true,
            srandmember: true,
            lindex: true,
            llen: true,
            lrange: true,
            dump: true,
            exists: true,
            keys: true,
            pfcount: true,
            hexists: true,
            hget: true,
            hgetall: true,
            hkeys: true,
            hlen: true,
            hstrlen: true,
            hvals: true,
            hscan: true,
            geohash: true,
            geopos: true,
            geodist: true,
            georadius: true,
            georadiusbymember: true
        }
    }

    pfcount(args, callback) {
        let aggregatedErrors = {}

        // Make keys buffer so we also get results in buffers. Required, because dump/restore fails with strings
        for (let i = 0; i < args[0].length; i++) {
            args[0][i] = Buffer.from(args[0][i])
        }

        // Dump all keys from all shards
        this._scatter("dump", args, this.writePolicy, (errors, results) => {
            if (errors && Object.keys(errors).length > 0) {
                aggregatedErrors["dump"] = errors
            }

            results = Helper.flatten(results, this.separator)

            // Get the one shard needed
            const baseKey = `${this.prefix}_${Date.now()}_${Math.random()}`
            const baseClients = Helper.getClients(this._clients, baseKey, this.replication)

            // Restore all keys to one shard
            this._restore(results, baseKey, baseClients, (errors, results, keys) => {
                if (errors && Object.keys(errors).length > 0) {
                    aggregatedErrors["restore"] = errors
                }

                // Do the required command
                if (keys.length > 0) {
                    this._send(baseClients, 0, true, "pfcount", [keys], (error, result) => {
                        if (error && Object.keys(error).length > 0) {
                            aggregatedErrors["pfcount"] = error
                        }

                        // Remove the gathered data
                        this._delete(keys, baseClients, (errors) => {
                            if (errors && Object.keys(errors).length > 0) {
                                aggregatedErrors["del"] = errors
                            }

                            aggregatedErrors = Helper.flatten(aggregatedErrors, this.separator)

                            if (Object.keys(aggregatedErrors).length === 0) {
                                aggregatedErrors = null
                            }

                            return callback(aggregatedErrors, result)
                        })
                    })
                } else {
                    aggregatedErrors = Helper.flatten(aggregatedErrors, this.separator)

                    if (Object.keys(aggregatedErrors).length === 0) {
                        aggregatedErrors = null
                    }

                    return callback(aggregatedErrors, null)
                }
            })
        })
    }

    mget(args, callback) {
        this._scatter("get", args, this.writePolicy, (errors, results) => {
            errors = Helper.flatten(errors, this.separator)

            if (Object.keys(errors).length === 0) {
                errors = null
            }

            results = Helper.flatten(results, this.separator)

            let preparedResults = {}

            if (this.debug === false) {
                for (let path in results) {
                    preparedResults[path.split(this.separator)[0]] = results[path]
                }
            } else {
                preparedResults = results
            }

            return callback(errors, preparedResults)
        })
    }

    _restore(keys, baseKey, clients, callback) {
        const errors = {}
        const results = {}

        const resultKeys = []
        let x = Object.keys(keys).length

        for (let key in keys) {
            if (keys[key]) {
                const tmpKey = `${baseKey}_${key}`

                resultKeys.push(tmpKey)

                let y = clients.length

                results[tmpKey] = {}

                for (let i = 0; i < clients.length; i++) {
                    this._send(clients, i, false, "restore", [tmpKey, 0, keys[key]], (error, result, clientId) => {
                        if (error) {
                            if (typeof errors[tmpKey] === "undefined") {
                                errors[tmpKey] = {}
                            }

                            errors[tmpKey][clients[clientId][0]] = error
                        }

                        results[tmpKey][clients[clientId][0]] = result

                        if (--y === 0) {
                            if (--x === 0) {
                                return callback(errors, results, resultKeys)
                            }
                        }
                    })
                }
            } else {
                if (--x === 0) {
                    return callback(errors, results, resultKeys)
                }
            }
        }
    }

    _delete(keys, clients, callback) {
        const errors = {}
        const results = {}

        let x = keys.length

        for (let i = 0; i < keys.length; i++) {
            results[keys[i]] = {}

            let y = clients.length

            for (let j = 0; j < clients.length; j++) {
                this._send(clients, j, false, "del", [keys[i]], (error, result, clientId) => {
                    if (error) {
                        if (typeof errors[keys[i]] === "undefined") {
                            errors[keys[i]] = {}
                        }

                        errors[keys[i]][clients[clientId][0]] = error
                    }

                    results[keys[i]][clients[clientId][0]] = result

                    if (--y === 0) {
                        if (--x === 0) {
                            return callback(errors, results)
                        }
                    }
                })
            }
        }
    }
}

module.exports = Commands

tools/benchmark.js
"use strict"

global.__base = __dirname + "/../"

const async = require("async")
const Benchmark = require("benchmark")
const redis = require("redis")

const Ared = require(`${__base}libs/index`)

const redisOptions = {
    r1: {
        host: "127.0.0.1",
        port: 6379,
        detect_buffers: true, // eslint-disable-line camelcase,
        enable_offline_queue: false // eslint-disable-line camelcase
    },
    r2: {
        host: "127.0.0.1",
        port: 6380,
        detect_buffers: true, // eslint-disable-line camelcase,
        enable_offline_queue: false // eslint-disable-line camelcase
    },
    r3: {
        host: "127.0.0.1",
        port: 6381,
        detect_buffers: true, // eslint-disable-line camelcase,
        enable_offline_queue: false // eslint-disable-line camelcase
    },
    r4: {
        host: "127.0.0.1",
        port: 6382,
        detect_buffers: true, // eslint-disable-line camelcase,
        enable_offline_queue: false // eslint-disable-line camelcase
    }
}

const keysPerSample = 16
const keys = []

let rawRedis = null
const basic1Ared = new Ared()
const basic2Ared = new Ared()
const basic4Ared = new Ared()
const basic4Ared2Replication2Write = new Ared()

basic4Ared2Replication2Write.replication = 4
basic4Ared2Replication2Write.writePolicy = 4

async.parallel([
    (callback) => {
        for (let id in redisOptions) {
            const client = redis.createClient(redisOptions[id])

            client.on("ready", () => {
                client.send_command("FLUSHALL")
            })
        }

        return callback()
    },

    (callback) => {
        for (let i = 0; i < keysPerSample; i++) {
            const key = Math.floor(Math.random() * 1000000000).toString()

            keys.push(key)
        }

        return callback()
    },

    (callback) => {
        rawRedis = redis.createClient(redisOptions["r1"])

        rawRedis.on("ready", () => {
            return callback()
        })
    },

    (callback) => {
        basic1Ared.listen(null, {r1: redisOptions["r1"]}, null, () => {
            return callback()
        })
    },

    (callback) => {
        basic2Ared.listen(null, redisOptions, null, () => {
            return callback()
        })
    },

    (callback) => {
        basic4Ared.listen(null, redisOptions, null, () => {
            return callback()
        })
    },

    (callback) => {
        basic4Ared2Replication2Write.listen(null, redisOptions, null, () => {
            return callback()
        })
    }
], () => {
    const options = {
        minSamples: 50,
        defer: true
    }

    const suite = new Benchmark.Suite()

    suite.add("Raw Redis Set", (deferred) => {
        let x = keysPerSample

        for (let i = 0; i < keysPerSample; i++) {
            rawRedis.send_command("set", [keys[i], keys[i]], () => {
                if (--x === 0) {
                    deferred.resolve()
                }
            })
        }
    }, options)

    suite.add("Raw Redis Get", (deferred) => {
        let x = keysPerSample

        for (let i = 0; i < keysPerSample; i++) {
            rawRedis.send_command("get", [keys[i]], () => {
                if (--x === 0) {
                    deferred.resolve()
                }
            })
        }
    }, options)

    suite.add("Basic 1 Redis Set", (deferred) => {
        let x = keysPerSample

        for (let i = 0; i < keysPerSample; i++) {
            basic1Ared.exec("set", [keys[i], keys[i]], () => {
                if (--x === 0) {
                    deferred.resolve()
                }
            })
        }
    }, options)

    suite.add("Basic 1 Redis Get", (deferred) => {
        let x = keysPerSample

        for (let i = 0; i < keysPerSample; i++) {
            basic1Ared.exec("get", [keys[i]], () => {
                if (--x === 0) {
                    deferred.resolve()
                }
            })
        }
    }, options)

    suite.add("Basic 2 Redis Set", (deferred) => {
        let x = keysPerSample

        for (let i = 0; i < keysPerSample; i++) {
            basic2Ared.exec("set", [keys[i], keys[i]], () => {
                if (--x === 0) {
                    deferred.resolve()
                }
            })
        }
    }, options)

    suite.add("Basic 2 Redis Get", (deferred) => {
        let x = keysPerSample

        for (let i = 0; i < keysPerSample; i++) {
            basic2Ared.exec("get", [keys[i]], () => {
                if (--x === 0) {
                    deferred.resolve()
                }
            })
        }
    }, options)

    suite.add("Basic 4 Redis Set", (deferred) => {
        let x = keysPerSample

        for (let i = 0; i < keysPerSample; i++) {
            basic4Ared.exec("set", [keys[i], keys[i]], () => {
                if (--x === 0) {
                    deferred.resolve()
                }
            })
        }
    }, options)

    suite.add("Basic 4 Redis Get", (deferred) => {
        let x = keysPerSample

        for (let i = 0; i < keysPerSample; i++) {
            basic4Ared.exec("get", [keys[i]], () => {
                if (--x === 0) {
                    deferred.resolve()
                }
            })
        }
    }, options)

    suite.add("Basic 4 Redis 2x Replication 2x Write Set", (deferred) => {
        let x = keysPerSample

        for (let i = 0; i < keysPerSample; i++) {
            basic4Ared2Replication2Write.exec("set", [keys[i], keys[i]], () => {
                if (--x === 0) {
                    deferred.resolve()
                }
            })
        }
    }, options)

    suite.add("Basic 4 Redis 2x Replication 2x Write Get", (deferred) => {
        let x = keysPerSample

        for (let i = 0; i < keysPerSample; i++) {
            basic4Ared2Replication2Write.exec("get", [keys[i]], () => {
                if (--x === 0) {
                    deferred.resolve()
                }
            })
        }
    }, options)

    suite.on("cycle", function (event) {
        console.log(String(event.target)) // eslint-disable-line no-console
    })

    suite.on("complete", function () {
        console.log("Fastest is " + this.filter("fastest").map("name")) // eslint-disable-line no-console

        process.exit(1)
    })

    suite.run({async: true})
})